src/probe_modules/packet.h

```diff
@@ -123,14 +158,78 @@ check_dst_port(uint16_t port, int num_ports, uint32_t *validation)
 	return (((max - min) % num_ports) >= ((to_validate - min) % num_ports));
 }
 
-static __attribute__((unused)) inline uint16_t
+static UNUSED inline uint16_t
 get_src_port(int num_ports, int probe_num, uint32_t *validation)
 {
 	return zconf.source_port_first +
 	       ((validation[1] + probe_num) % num_ports);
 }
 
+static inline struct ip *get_ip_header(const u_char *packet, uint32_t len)
+{
+	// buf not large enough to contain expected udp header
+	if (len < sizeof(struct ether_header)) {
+		return NULL;
+	}
+	return (struct ip *)&packet[sizeof(struct ether_header)];
+}
+
+static inline struct tcphdr *get_tcp_header(const struct ip *ip_hdr,
+					    uint32_t len)
+{
+	// buf not large enough to contain expected udp header
+	if ((4 * ip_hdr->ip_hl + sizeof(struct tcphdr)) > len) {
+		return NULL;
+	}
+	return (struct tcphdr *)((char *)ip_hdr + 4 * ip_hdr->ip_hl);
+}
+
+static inline struct udphdr *get_udp_header(const struct ip *ip_hdr,
+					    uint32_t len)
+{
+	// buf not large enough to contain expected udp header
+	if ((4 * ip_hdr->ip_hl + sizeof(struct udphdr)) > len) {
+		return NULL;
+	}
+	return (struct udphdr *)((char *)ip_hdr + 4 * ip_hdr->ip_hl);
+}
+
+static inline struct icmp *get_icmp_header(const struct ip *ip_hdr,
+					   uint32_t len)
+{
+	// buf not large enough to contain expected udp header
+	if ((4 * ip_hdr->ip_hl + sizeof(struct icmp)) > len) {
+		return NULL;
+	}
+	return (struct icmp *)((char *)ip_hdr + 4 * ip_hdr->ip_hl);
+}
+
+static inline char *get_udp_payload(const struct udphdr *udp,
```

Comment:
I'd do uint8_t instead of char, since we think of it as a byte array, not a string.